server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,412 @@
+ * caller will be accessing segment files in the same way as {@code FSDirectory}. Apart from storing actual segment files,
+ * remote segment store also keeps track of refresh and commit checkpoints in a separate path which is handled by
+ * another instance of {@code RemoteDirectory}.
+ * @opensearch.internal
+ */
+public final class RemoteSegmentStoreDirectory extends FilterDirectory {
+    /**
+     * Each segment file is uploaded with unique suffix.
+     * For example, _0.cfe in local filesystem will be uploaded to remote segment store as _0.cfe__gX7bNIIBrs0AUNsR2yEG
+     */
+    public static final String SEGMENT_NAME_UUID_SEPARATOR = "__";
+
+    /**
+     * remoteDataDirectory is used to store segment files at path: cluster_UUID/index_UUID/shardId/segments/data
+     */
+    private final RemoteDirectory remoteDataDirectory;
+    /**
+     * remoteMetadataDirectory is used to store metadata files at path: cluster_UUID/index_UUID/shardId/segments/metadata
+     */
+    private final RemoteDirectory remoteMetadataDirectory;
+
+    /**
+     * To prevent explosion of refresh metadata files, we replace refresh files for the given primary term and generation
+     * This is achieved by uploading refresh metadata file with the same suffix as last commit metadata file.
+     */
+    private String refreshMetadataFileUniqueSuffix;
+    /**
+     * Keeps track of local segment filename to uploaded filename along with other attributes like checksum.
+     * This map acts as a cache layer for uploaded segment filenames which helps avoid calling listAll() each time.
+     * It is important to initialize this map on creation of RemoteSegmentStoreDirectory and update it on each upload and delete.
+     */
+    private Map<String, UploadedSegmentMetadata> segmentsUploadedToRemoteStore;
+
+    private static final Logger logger = LogManager.getLogger(RemoteSegmentStoreDirectory.class);
+
+    public RemoteSegmentStoreDirectory(RemoteDirectory remoteDataDirectory, RemoteDirectory remoteMetadataDirectory) throws IOException {
+        super(remoteDataDirectory);
+        this.remoteDataDirectory = remoteDataDirectory;
+        this.remoteMetadataDirectory = remoteMetadataDirectory;
+        init();
+    }
+
+    /**
+     * Initializes the cache which keeps track of all the segment files uploaded to the remote segment store.
+     * As this cache is specific to an instance of RemoteSegmentStoreDirectory, it is possible that cache becomes stale
+     * if another instance of RemoteSegmentStoreDirectory is used to upload/delete segment files.
+     * It is caller's responsibility to call init() again to ensure that cache is properly updated.
+     * @throws IOException if there were any failures in reading the metadata file
+     */
+    public void init() throws IOException {
+        this.refreshMetadataFileUniqueSuffix = UUIDs.base64UUID();
+        this.segmentsUploadedToRemoteStore = new ConcurrentHashMap<>(readLatestMetadataFile());
```

Comment:
We haven't really uploaded any segments to remote store at this point right? or does this handle the cases where replica gets promoted to primary?

Reply:
Yes, failover or node restart.

Reply:
@sachinpkale it seems like there is no cleanup mechanism in place: if init() is called for initialized bu stale cache, the stale files have to be cleaned up (using previous `metadataFileUniqueSuffix`).

Reply:
@reta We always read latest metadata file. So, even if there are more than 1 metadata files, it won't impact the correctness.
But you are right on cleanup part. We need cleanup of stale metadata files in order to keep performance of `init()` in check. I will be raising a follow-up PR which will contain clean-up logic. Implementation is part of another branch: https://github.com/sachinpkale/OpenSearch/commit/b0d95c9534e10460605167ce29e9018ea5495ff6#diff-f57628264c235a133d2d6db458404e07a43f18026e40ed5e956536c03add9925R273